third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor_v2.rs

```diff
@@ -0,0 +1,1883 @@
+    /// locals being released need to be determined from livevar analysis results.
+    pub fn released_temps(&self) -> impl Iterator<Item = TempIndex> + '_ {
+        self.before
+            .local_to_label_map
+            .keys()
+            .filter(|t| !self.after.local_to_label_map.contains_key(t))
+            .cloned()
+    }
+
+    /// Returns true if the value in the variable has been moved at this program point.
+    pub fn is_moved(&self, temp: TempIndex) -> bool {
+        self.after.moved.contains(&temp)
+    }
+}
+
+impl FunctionTargetProcessor for ReferenceSafetyProcessor {
+    fn process(
+        &self,
+        _targets: &mut FunctionTargetsHolder,
+        fun_env: &FunctionEnv,
+        mut data: FunctionData,
+        _scc_opt: Option<&[FunctionEnv]>,
+    ) -> FunctionData {
+        if fun_env.is_native() {
+            return data;
+        }
+        let target = FunctionTarget::new(fun_env, &data);
+        let live_var_annotation = target
+            .get_annotations()
+            .get::<LiveVarAnnotation>()
+            .expect("livevar annotation");
+        let suppress_errors = fun_env
+            .module_env
+            .env
+            .get_extension::<Options>()
+            .unwrap_or_default()
+            .experiment_on(Experiment::NO_SAFETY);
+        let analyzer = LifeTimeAnalysis {
+            target: &target,
+            live_var_annotation,
+            suppress_errors,
+        };
+        let code = target.get_bytecode();
+        let cfg = StacklessControlFlowGraph::new_forward(code);
+        let mut state = LifetimeState::default();
+        let mut label_counter: u32 = 0;
+        for (i, Parameter(_, ty, loc)) in fun_env.get_parameters().into_iter().enumerate() {
+            if ty.is_reference() {
+                let label = LifetimeLabel::new_from_counter(label_counter);
+                label_counter += 1;
+                state.new_node(label, MemoryLocation::External);
+                let target = state.make_local_from_label_fun(
+                    i,
+                    || LifetimeLabel::new_from_counter(label_counter),
+                    false,
+                );
+                label_counter += 1;
+                state.add_edge(label, BorrowEdge {
+                    kind: BorrowEdgeKind::BorrowLocal(ty.is_mutable_reference()),
+                    loc,
+                    target,
+                })
+            }
+        }
+        let state_map = analyzer.analyze_function(state, target.get_bytecode(), &cfg);
+        let mut state_map_per_instr = analyzer.state_per_instruction(
+            state_map,
+            target.get_bytecode(),
+            &cfg,
+            |before, after| LifetimeInfoAtCodeOffset {
+                before: before.clone(),
+                after: after.clone(),
+            },
+        );
+        // For dead code, there may be holes in the map. Identify those and populate with default so
```

Comment:
*Not* an action item for this PR, but I wonder if it would be a good to move the generic version of this code into `state_per_instruction` method?